torchx/mix.exs

```diff
@@ -43,4 +44,145 @@ defmodule Torchx.MixProject do
       source_url: @source_url
     ]
   end
+
+  defp aliases do
+    [
+      "compile.torchx": &compile/1
+    ]
+  end
+
+  defp compile(args) do
+    cache_dir = Path.join(__DIR__, "cache")
+
+    force_download = "--force" in args
+
+    libtorch_dir =
+      case System.get_env("LIBTORCH_DIR") do
+        nil ->
+          Path.join(cache_dir, "libtorch")
+
+        val ->
+          val
+      end
+
+    with {:force_download, false} <- {:force_download, force_download},
+         {:libtorch_dir_exists, true} <- {:libtorch_dir_exists, File.dir?(libtorch_dir)},
+         {:ok, content} <- File.read(Path.join(libtorch_dir, "build-hash")),
+         true <- String.contains?(content, "pytorch") do
+      :noop
+    else
+      {:force_download, true} ->
+        File.rm_rf!(cache_dir)
+        install_libtorch(cache_dir, libtorch_dir)
+
+      {:libtorch_dir_exists, false} ->
+        install_libtorch(cache_dir, libtorch_dir)
+
+      _ ->
+        msg = "directory #{libtorch_dir} set in LIBTORCH_DIR does not contain libtorch"
+        IO.puts(msg)
+
+        {:error,
+         [
+           %Mix.Task.Compiler.Diagnostic{
+             compiler_name: "torchx",
+             details: msg,
+             file: "mix.exs",
+             message: msg,
+             position: nil,
+             severity: :error
+           }
+         ]}
+    end
+  end
+
+  defp install_libtorch(cache_dir, libtorch_dir) do
+    File.mkdir_p!(cache_dir)
+
+    libtorch_zip = System.get_env("LIBTORCH_ZIP", Path.join(cache_dir, "libtorch.zip"))
+
+    os =
+      case System.cmd("uname", ~w(-s)) do
+        {"Linux" <> _, 0} ->
+          :linux
+
+        {"Darwin" <> _, 0} ->
+          :mac
+
+        {os, 0} ->
+          raise "OS #{os} is not supported"
+
+        _ ->
+          raise "unable to fetch current OS"
+      end
+
+    unless File.exists?(libtorch_zip) do
+      # Download libtorch
+
+      url =
+        case os do
+          :linux ->
+            "https://download.pytorch.org/libtorch/cpu/libtorch-cxx11-abi-shared-with-deps-1.9.1%2Bcpu.zip"
+
+          :mac ->
+            "https://download.pytorch.org/libtorch/cpu/libtorch-macos-1.9.1.zip"
+        end
+
+      download!(url, libtorch_zip)
+    end
+
+    # Unpack libtorch to libtorch_dir
+
+    {_, 0} =
+      System.cmd("unzip", ~w(-d #{cache_dir} #{libtorch_zip})) |> IO.inspect(label: "unzip")
+
+    unpack_dir = Path.join(cache_dir, "libtorch") |> IO.inspect(label: "unpack dir")
+
+    unless unpack_dir == libtorch_dir do
+      {_, 0} = System.cmd("mv", ~w(#{unpack_dir} #{libtorch_dir}))
```

Comment:
You can use `File.mv!` :)

Reply:
Huh, I didn't find it in the autocomplete

Reply:
```elixir
iex(1)> File.     
CopyError        Error            LinkError        RenameError      
Stat             Stream           cd!/1            cd!/2            
cd/1             chgrp!/2         chgrp/2          chmod!/2         
chmod/2          chown!/2         chown/2          close/1          
copy!/2          copy!/3          copy/2           copy/3           
cp!/2            cp!/3            cp/2             cp/3             
cp_r!/2          cp_r!/3          cp_r/2           cp_r/3           
cwd!/0           cwd/0            dir?/1           dir?/2           
exists?/1        exists?/2        ln!/2            ln/2             
ln_s!/2          ln_s/2           ls!/0            ls!/1            
ls/0             ls/1             lstat!/1         lstat!/2         
lstat/1          lstat/2          mkdir!/1         mkdir/1          
mkdir_p!/1       mkdir_p/1        open!/1          open!/2          
open!/3          open/1           open/2           open/3           
read!/1          read/1           read_link!/1     read_link/1      
regular?/1       regular?/2       rename!/2        rename/2         
rm!/1            rm/1             rm_rf!/1         rm_rf/1          
rmdir!/1         rmdir/1          stat!/1          stat!/2          
stat/1           stat/2           stream!/1        stream!/2        
stream!/3        touch!/1         touch!/2         touch/1          
touch/2          write!/2         write!/3         write/2          
write/3          write_stat!/2    write_stat!/3    write_stat/2     
write_stat/3  
```

This is on Elixir 1.12.2

Reply:
Ah sorry, it's `rename`!

Reply:
Nice!